File: backend/app/api/runs.py
"""Run execution endpoints."""

from fastapi import APIRouter, HTTPException, status, BackgroundTasks

from backend.app.core.pagination import apply_pagination_and_sort, build_paginated_response
from backend.app.schemas.run import RunCreate, RunResponse, RunDetailResponse, PaginatedRunResponse
from backend.app.core.dependencies import DatabaseSession, CurrentUser
from backend.app.services.run_service import create_run
from backend.app.services.runner import run_executor
from backend.app.models.run import Run


# 新增
from fastapi import Query
from sqlalchemy import select, func
from backend.app.models.ticket import Ticket


router = APIRouter(prefix="/runs", tags=["Runs"])


@router.post("", response_model=RunResponse)
async def create_new_run(
    run_in: RunCreate,
    background_tasks: BackgroundTasks,
    db: DatabaseSession,
    current_user: CurrentUser
):
    """
    Create and execute a new run.

    The run will be executed in the background.
    """
    # Create run
    run = await create_run(db, run_in, current_user)

    # Execute run in background
    background_tasks.add_task(run_executor.execute_run, db, run)

    return run


# 修改
@router.get("", response_model=PaginatedRunResponse)
async def list_runs(
    db: DatabaseSession,
    current_user: CurrentUser,
    ticket_id: int | None = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    order_by: str = Query(default="created_at"),  # 【新增】
    order: str = Query(default="desc")            # 【新增】
):
    """
    List runs with pagination.

    Optionally filter by ticket_id.
    """
    query = select(Run)

    if ticket_id:
        query = query.where(Run.ticket_id == ticket_id)

    if not current_user.is_admin:
        query = query.join(Ticket).where(Ticket.created_by_id == current_user.id)

    count_result = await db.execute(select(func.count()).select_from(query.subquery()))
    total = count_result.scalar()

    result = await db.execute(
        apply_pagination_and_sort(query, Run, page, page_size, order_by, order)
    )
    runs = result.scalars().all()
    return build_paginated_response(runs, RunResponse, total, page, page_size)



@router.get("/{run_id}", response_model=RunDetailResponse)
async def get_run(
    run_id: int,
    db: DatabaseSession,
    current_user: CurrentUser
):
    """
    Get run details including logs.
    """
    result = await db.execute(select(Run).where(Run.id == run_id))
    run = result.scalar_one_or_none()

    if not run:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Run not found"
        )

    # Check permissions
    if not current_user.is_admin:
        from backend.app.models.ticket import Ticket
        ticket_result = await db.execute(select(Ticket).where(Ticket.id == run.ticket_id))
        ticket = ticket_result.scalar_one_or_none()
        if ticket.created_by_id != current_user.id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized to view this run"
            )

    return run

File: frontend/src/api/runs.ts
import request from "@/utils/request";
import type { Run } from "@/types/run";

// 查询参数类型
interface GetRunsParams {
  page?: number;      // 当前页码
  page_size?: number; // 每页条数
  ticket_id?: number; // 按工单 ID 筛选（可选）
  order_by?: string;       // 【新增】
  order?: "asc" | "desc"; // 【新增】
}

// 创建 Run 时需要传入的数据类型
interface CreateRunData {
  run_type: string;                       // 运行类型，如 "manual" / "auto" 等
  script_id: string;                      // 要执行的脚本 ID
  execution_context: Record<string, any>; // 执行上下文，键值对形式的动态参数
  ticket_id: number;                      // 关联的工单 ID
}

// 分页返回结构
interface PaginatedRuns {
  data: Run[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// 获取运行记录列表，支持分页和按工单筛选
export const getRuns = (params?: GetRunsParams) =>
  request.get<PaginatedRuns>("/runs", { params });

// 根据 ID 获取单条运行记录详情
export const getRun = (id: number) =>
  request.get<Run>(`/runs/${id}`);

// 创建一条新的运行记录（即触发一次脚本执行）
export const createRun = (data: CreateRunData) =>
  request.post<Run>("/runs", data);